### PR TITLE
Try using a template part in the navigation block

### LINKED ISSUE #34612: Navigation Block: save data to a custom post type
## What problem does this address?
One of the aspects of the navigation block discussed in the past has been how it saves its data.

In the past WordPress menus championed reusability, you can assign one menu to multiple locations on a site. When the navigation block was first envisaged, there was a discussion about whether the menu storage system (custom post types for menu items and I believe a 'term' for the menu itself) could be used, but it was decided that a more block-y approach should be taken, preferring the raw HTML output of blocks.

Since then, the way reusable blocks and template parts work has come along way, and it might be time to revisit whether the navigation block could use the same sort of system. This would essentially allow the user to create a menu once and reuse it anywhere on their site, with the blocks potentially being editable from a standalone editor

## What is your proposed solution?
Use a similar implementation as template parts and reusable blocks for the navigation block.

## Technical complications
I did a quick assessment of what would be required.

Reusable blocks and template parts are wrappers that save all of their inner blocks to a custom post type. The main difference to the navigation block is that the navigation block itself renders a `<nav>` wrapper and has a considerable number of attributes that also need to be saved and loaded, and from what I can tell, this is a new area yet to be technically explored.

An option would be to create a new wrapping block, but it would have to be interface-less.

## Opportunities
The way the 'theme opt-in' in the navigation editor works still needs to be shaped considerably (https://github.com/WordPress/gutenberg/issues/33969). There's a potential opportunity that when a theme opts-in to support the navigation block in the nav editor, the editor saves and loads the same custom post type system as the block.

### PR DESCRIPTION
**Pull request by talldan**
## Description
Closes #34612

This is a very unrefined PR that adds template part storage to the nav block.

Current problems:
- Navigation block template parts show up in the template part block and vice versa
- The navigation block now has a multi stage creation process that could use some refinement
- A lot of code is borrowed from the template part block, so the nav block looks a lot like a template part
- It'd be good to make sure existing (non-template part) navigation blocks can be saved in template parts.
- Navigation editor no longer works

## How has this been tested?
1. Open the site editor (avoid testing in the post editor because it doesn't seem to save template part entities)
2. Create a nav block

## Types of changes

New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/block-library/src/navigation/block.json [hunk 1]
```diff
@@ -4,9 +4,22 @@
...
+		"menu",
+		"navigation",
+		"links"
+	],
 	"textdomain": "default",
 	"attributes": {
+		"slug": {
+			"type": "string"
+		},
+		"theme": {
```
Comment (mtias) [inline packages/block-library/src/navigation/block.json:17]: Why would we need to keep a reference to "theme"?

Comment (talldan) [reply packages/block-library/src/navigation/block.json:17]: The PR is very copy/paste from the template part block at the moment. I'll look at removing this 👍

Comment (mtias) [reply packages/block-library/src/navigation/block.json:17]: Yes, makes sense :D I think we should be as lean as possible when it comes to borrowing from the template part structure.

**Review by getdave — COMMENTED** [review]
So this is very cool. Thanks so much for exploring it.

I gave it a test drive and the link to the video below has some commentary which explains what I found.

🎥 [Test Drive Video](https://d.pr/v/FM6yVN) (it was too large for Github!)

## TL;DR

* The basic process works well  - the "data" (ie: the items in your Nav) are saved to template part without the wrapping `core/navigation` block.
* You can reuse the same navigation in _different_ places on the same site and updates in one location will be reflected in another.
* You can access the navigation in the Isolated Template Part Editor and you will see only the "inner" navigation block items (no Navigation block).

Things we will need to address (I know you are aware of many of these):

* When in Isolated Template Part editor you cannot add child navigation-type blocks (e.g. `core/navigation-link` because there is no wrapping parent Navigation block and these blocks are set to only appear if they are a child of that block.
* I couldn't seem to access the nav template part on another Theme. This might be a limitation of the Template Parts themselves at the moment. The good thing is we will inherit much of the improvements if we reuse some/all of the Template Part mechanics.
* UI/UX - this obviously needs refinement to mask the "Template Part" specific UI and tailor towards Navigation. I think the user shouldn't necessarily need to know we are using Template Parts.

Overall however this is a great start and everyone should experiment with it to get a sense of where it could take us.

**Comment by noisysocks** [comment]
Very cool! Surprising to see it work so well at this early stage.

> * When in Isolated Template Part editor you cannot add child navigation-type blocks (e.g. `core/navigation-link` because there is no wrapping parent Navigation block and these blocks are set to only appear if they are a child of that block.

This exposes an interesting question. Do we want the settings which are stored on the `core/navigation` block to be shared across templates? For example, if the navigation has a horizontal layout then would we want it to be appear horizontal in all other templates where it's used?

If yes, then I think the solution here is to store the `core/navigation` block _inside_ the template part CPT. With this approach there isn't really a difference between a navigation and a template part. Perhaps `core/navigation` can be hidden from the inserter and "Navigation" is just a variation of the `core/template-part` block that contains a `core/navigation`.

If not, then I think the solution here is to use a different CPT for navigation so that it doesn't appear in the template part editor. We can build a custom template part editor for navigation which provides the wrapping `core/navigation` block.

**Comment by talldan** [comment]
> If not, then I think the solution here is to use a different CPT for navigation so that it doesn't appear in the template part editor. We can build a custom template part editor for navigation which provides the wrapping core/navigation block.

The goal right now is to keep the inner blocks separate from the settings as you mention here, the idea being that menu data can be reused but given a different appearance depending on the location.

The navigation block also has the inverse problem that you shouldn't be able to choose non navigation block template parts from the placeholder. Either the template part could have a way to identify the block type it's associated with, or as you say a separate CPT could be used. Some previous discussion here - https://github.com/WordPress/gutenberg/issues/34612#issuecomment-932160213.

**Comment by getdave** [comment]
Good questions @noisysocks!

> Do we want the settings which are stored on the `core/navigation` block to be shared across templates? For example, if the navigation has a horizontal layout then would we want it to be appear horizontal in all other templates where it's used?

I think we wouldn't want this. For example if I have a primary menu I want to display that in the header and the footer but in totally different ways.

For example, the header might be horizontal with dropdown menus. The footer might be horizontal but with all submenus shown vertical below each item (like a "fat footer"). The data should be the same but the styling/layout might need to be completely different.

IMHO the layout/appearance rules applied to the Nav block itself should be specific to the insertion location in the Template where it is added. That means _not_ storing it alongside the inner blocks within the post entity. The keeps styling/layout (Nav block as "Template Part") separate from "data" (the navigation items as inner blocks). 

This also means we could reuse "just the data" in the the Navigation _Editor_ without having to undo/disable a load of layout/styling config coming from the Nav block itself. 

@talldan Does the above sound like what you are thinking too?

**Comment by adamziel** [comment]
Great work @talldan, keep it coming!
 
> Do we want the settings which are stored on the core/navigation block to be shared across templates? For example, if the navigation has a horizontal layout then would we want it to be appear horizontal in all other templates where it's used?

I actually like being able to have it vertical on one page and horizontal on another. If I want to share that, nothing stops me from creating a new wrapping template part just for the Navigation block.

> This also means we could reuse "just the data" in the the Navigation Editor without having to undo/disable a load of layout/styling config coming from the Nav block itself.

I'm all for that, but I can't think of an approach that doesn't involve duplicating a lot of the code from the nav block. Exporting `NavigationInnerBlocks` from block-library would help, but seems controversial. Any ideas?

**Comment by getdave** [comment]
> I'm all for that, but I can't think of an approach that doesn't involve duplicating a lot of the code from the nav block

I assume you mean CSS and also some of the JS utilities that come baked into the Nav block?

I starting to think perhaps having the CSS of the Nav block decoupled from the editor is a good thing. If the inner blocks' styles are similarly decoupled from the parent Nav block then this should mean they will "still work" with just a little custom CSS in the Nav _Editor_.

In terms of utilities I assume we could look to share a lot of this via some kind of shared package such as `@wordpress/menus`?

Just initial thoughts - I haven't dived into any of this yet.

**Comment by adamziel** [comment]
> > I'm all for that, but I can't think of an approach that doesn't involve duplicating a lot of the code from the nav block
> 
> I assume you mean CSS and also some of the JS utilities that come baked into the Nav block?

I meant all the JS from the Navigation block and NavigationInnerBlocks component, there is a fair bit of work done there.

> I starting to think perhaps having the CSS of the Nav block decoupled from the editor is a good thing. If the inner blocks' styles are similarly decoupled from the parent Nav block then this should mean they will "still work" with just a little custom CSS in the Nav _Editor_.

👍 

> In terms of utilities I assume we could look to share a lot of this via some kind of shared package such as `@wordpress/menus`?

It would work, although `@wordpress/block-library` would have to depend on it – if that's not an issue then we may be good.

**Comment by draganescu** [comment]
Aww 😍 we're moving to the place!

> I think the solution here is to store the `core/navigation` block inside the template part CPT 

We'd be back to having to configure the block from outside settings and we'd also step away from the "reuse data" idea.

I think we should just not make it a "template part" but a  "navigation" .... part, case in which we can tweak the behavior a lot: auto wrap in a nav block, correct what blocks are available to insert and so on.

**Comment by talldan** [comment]
> @talldan Does the above sound like what you are thinking too?

I think you might have missed the comment I made just before yours commenting on that question, might have been a simultaneous post - https://github.com/WordPress/gutenberg/pull/35418#issuecomment-938299395

It seems like there are a lot of good advantages from not saving the navigation block wrapper in the template part.

It does mean wherever that template part is edited in isolation (navigation editor, template part editor, and potentially the template part focus mode) has some complexity, similar to what we've already been experiencing in the navigation editor. 

If left as it is now, we'd have a situation where only the inner blocks can be edited. This breaks some concepts like how the Navigation Link has a 'parent' property. The 'direct insert' behavior that was just implemented no longer works and would need to be replicated.

The other option is to include a navigation block as a locked wrapper in these isolated editors. This fixes the things that were just mentioned, but also has some drawbacks. All the styling options on the navigation block need to be disabled (except for in template part focus mode, where this might still be relevant). Another conundrum is that the navigation block will likely have an option to switch to  a different template part, but does that make sense when editing a single template part already? It would seem like that's the job of the editor.

**Comment by tellthemachines** [comment]
This looks very promising 🚀 

I think conceptually at least it makes sense for Navigation to be its own custom post type, because template parts by definition are freeform, all-blocks-allowed areas, whereas what we want here is a block that is constrained to accept only certain types of blocks, and has a bunch of custom behaviour (direct insert and so forth). 

It would also mean we could have a "Navigation" section in the Appearance menu, which would make it easier to manage menus in the admin. As it is, we need to go looking for them in the Template Parts page. And perhaps, when opening a Navigation CPT from the Appearance menu, we could show it in a navigation editor type interface 😄 

Would creating a new CPT imply a lot of duplication of logic that we already have in the Template Part?

**Comment by talldan** [comment]
> Would creating a new CPT imply a lot of duplication of logic that we already have in the Template Part?

That's definitely the concern.

File: packages/block-library/src/navigation/block.json (hunk 1 above)
Comment (talldan) [reply packages/block-library/src/navigation/block.json:17]: Yet to find an easy way to achieve this, it seems as though a template part can only be fetched when it has a `theme`.

**Comment by adamziel** [comment]
> Would creating a new CPT imply a lot of duplication of logic that we already have in the Template Part?

How does a new CPT compare to having template parts "variations" with limited capabilities e.g. configurable `allowedBlocks`? I'm thinking of "general" template part and "navigation" template part. Are there any other use cases for that? A "post" template part perhaps?

File: packages/block-library/src/navigation/block.json (hunk 1 above)
Comment (mtias) [reply packages/block-library/src/navigation/block.json:17]: We should probably remove that limitation, it's coupling template parts too much with themes, and it should be entirely feasible (maybe even the norm down the road) to have template parts that are not associated with any specific theme at all.

**Comment by mtias** [comment]
Thank you very much for diving into this exploration. It seems the most important decision is whether the template part abstraction is a good enough fit, and provides enough value that it offsets the extra work needed to accommodate the specificities and differences that will inevitably arise between navigation and general purpose template parts.

Some of these differences might be pointing at shortcomings of the template part structure, which we should improve (for example, how tied things are to a `theme` when it comes to template parts, which I think is shortsighted on our design). Others might end up with so many exclusions (like not wanting it show in the "mosaic view" of template parts, or the template part selectors, etc) that we'd end up spreading through the codebase exclusions for navigation and a maintenance burden.

We should probably list those considerations, both concrete benefits we get and negatives we incur when using `wp_template_part` vs a new `wp_navigation` modelled after it.

**Comment by talldan** [comment]
> How does a new CPT compare to having template parts "variations" with limited capabilities e.g. configurable allowedBlocks? I'm thinking of "general" template part and "navigation" template part. Are there any other use cases for that? A "post" template part perhaps?

Template parts do already use variations for semantic areas (Header, Footer).

**Comment by talldan** [comment]
> We should probably list those considerations, both concrete benefits we get and negatives we incur when using wp_template_part vs a new wp_navigation modelled after it.

Yep, agreed. Here's an attempt to define some of the problems we're currently facing
- A way to associate a block type with a template part might be needed. This would prevent other non-navigation template parts being visible in the navigation block (and also the opposite).
- Navigation block template parts are also editable from wp-admin and the focus mode, but many of the inner block types of the navigation block have a 'parent' defined and can only be inserted within a particular parent. Editing in isolation will likely need a special case to either wrap blocks with a navigation block or for the editor to imply the inner blocks are within a navigation block without one actually being present.
- Template parts seem to be strongly associated with a theme and a template part created in one theme is unavailable in another. This isn't a requirement for the navigation block. Is it a requirement for custom template parts, or is this something that needs to be fixed?
- Template parts have an 'area' that defines the semantic region of a page. From what I can tell this doesn't really fit in with the way a navigation block might have a location. For a template part, the area **defines** a semantic region of a site (e.g. header). For a navigation block, the location is defined by where it is rendered. A single navigation block template part might also be rendered in more than one place, so a single area doesn't make sense. It might be interesting to look at a way for the navigation block to inherit a location from a parent template part's area. The parent area could also be useful for suggesting patterns or styles similar to how a template part does.
- Also worth mentioning the current use cases for a navigation block 'location'. Theme switching is one, where a classic menu location might have to be written to a template part. The API described in #30674 might be another if a user wants to build a sitemap based on their main menu.

**Comment by mtias** [comment]
> This would prevent other non-navigation template parts being visible in the navigation block (and also the opposite)

This would be relatively straightforward from the perspective of the navigation block, and would follow how specific areas can be swapped. Here you only see "headers" for example and not other template parts:

image

I think the challenge is more about the cases where all template parts are shown, and whether we need to exclude navigation from those. That's what could start to get things a bit convoluted in my view.

> Template parts seem to be strongly associated with a theme and a template part created in one theme is unavailable in another. This isn't a requirement for the navigation block. Is it a requirement for custom template parts, or is this something that needs to be fixed?

The latter. Currently the coupling is overbearing and a self-imposed limitation. I should be able to access template parts I've saved on my site regardless of what theme I'm using. The theme a template part was _created in_ can be useful information but should not be used as a strong filter.

> A single navigation block template part might also be rendered in more than one place, so a single area doesn't make sense.

The area in this case would also loosely refer to `nav`. It's possible to have more than one header in a page, etc.

> Navigation block template parts are also editable from wp-admin and the focus mode, but many of the inner block types of the navigation block have a 'parent' defined and can only be inserted within a particular parent.

I think the `parent` here should be equivalent and somehow present in the isolated template part as a provider. Also somewhat related: #30641.

**Comment by carlomanf** [comment]
> Template parts seem to be strongly associated with a theme and a template part created in one theme is unavailable in another. This isn't a requirement for the navigation block. Is it a requirement for custom template parts, or is this something that needs to be fixed?

#31971 goes a long way towards fixing this.

**Comment by talldan** [comment]
I've made an alternative PR that uses a `wp_navigation` post type - https://github.com/WordPress/gutenberg/pull/35746

**Comment by talldan** [comment]
> I think the challenge is more about the cases where all template parts are shown, and whether we need to exclude navigation from those. That's what could start to get things a bit convoluted in my view.

@mtias Yes, I agree with this, it's definitely where things are getting complicated. If there were a way to solve this a lot of the other problems would also be easier to solve, but I question if it's something that should be solved. Feels like it would be introducing lots of granularity to template parts in order to reuse some of the existing systems like areas. Instead, maybe some of those systems should be abstracted to work more easily with different post types.

As mentioned above, I've made #35746, which naturally provides a separate page for listing menus:
Screenshot 2021-10-19 at 12 17 55 pm

But we still have the same issue that the post editor isn't currently suitable for editing these menus. I guess this is where we see the 'Isolated Editor' being used? Should this listing even be available in the admin?

> I should be able to access template parts I've saved on my site regardless of what theme I'm using.

Looks like @adamziel has created a PR to solve this (#35666) 🎉 

> The area in this case would also loosely refer to nav. It's possible to have more than one header in a page, etc.

Would you be able to elaborate on 'loosely refer to nav'?

> I think the parent here should be equivalent and somehow present in the isolated template part as a provider. Also somewhat related: #30641.

Yeah, that would be interesting. In the Navigation (and to some degree Widgets) editors, hardcoded locked templates were used, but that approach wasn't perfect. It feels like the block should instead be assuming the role of the root block list.
